### PR TITLE
feat: First version of shoot comparison app

### PR DESCRIPTION
**Pull request by akgalwas**
**Description**

Changes proposed in this pull request:

- Introduced a CLI tool for comparing shoots produced by the Provisioner, and the KIM
- The cli allows comparing shoot files

**Related issue(s)**
#185

### CODE REVIEW

**Review by Disper — COMMENTED** [review]
@akgalwas have you considered placing the tool in the `hack` repository instead of `tools`?

**Review by Disper — CHANGES_REQUESTED** [review]
1. @akgalwas have you considered placing the tool in the `hack` repository instead of `tools`? 
2. Please create a README.md that will explain briefly the purpose of the tool and show an example usage.

**Comment by Disper** [comment]
I have changed three fields
- `spec.CloudProfileName`
- `spec.shootIssuer.enabled`
- `spec.provider.infrastructureConfig.networks.cidr`
 and compared the files.
 
 I expected the output to say that the 3 fields are different and show the difference. Instead, the tool said that just `spec` are different and printed the whole content. It was also too big so it was truncated. Is it intentional?

```
go run main.go files ~/Desktop/garden-kyma-dev-md-pvc-prov5.yaml ~/Desktop/prov-modified.yaml
Comparing files: /Users/i316752/Desktop/garden-kyma-dev-md-pvc-prov5.yaml and /Users/i316752/Desktop/prov-modified.yaml
Shoot files are NOT equal
spec: Expected
    <v1beta1.ShootSpec>: {
        Addons: nil,
        CloudProfileName: "az-modified",
        DNS: nil,
        Extensions: [
            {
                Type: "shoot-dns-service",
                ProviderConfig: {
                    Raw: "{\"apiVersion\":\"service.dns.extensions.gardener.cloud/v1alpha1\",\"dnsProviderReplication\":{\"enabled\":true},\"kind\":\"DNSConfig\"}",
                    Object: nil,
                },
                Disabled: nil,
            },
            {
                Type: "shoot-cert-service",
                ProviderConfig: {
                    Raw: "{\"apiVersion\":\"service.cert.extensions.gardener.cloud/v1alpha1\",\"kind\":\"CertConfig\",\"shootIssuers\":{\"enabled\":false}}",
                    Object: nil,
                },
                Disabled: nil,
            },
            {
                Type: "shoot-networking-filter",
                ProviderConfig: nil,
                Disabled: true,
            },
        ],
        Hibernation: nil,
        Kubernetes: {
            ClusterAutoscaler: nil,
            KubeAPIServer: {
                KubernetesConfig: {FeatureGates: nil},
                AdmissionPlugins: nil,
                APIAudiences: nil,
                AuditConfig: {
                    AuditPolicy: {
                        ConfigMapRef: {Kind: "", Namespace: "", Name: "audit-policy", UID: "", APIVersion: "", ResourceVersion: "", FieldPath: ""},
                    },
                },
                OIDCConfig: nil,
                RuntimeConfig: nil,
                ServiceAccountConfig: nil,
                WatchCacheSizes: nil,
                Requests: nil,
                EnableAnonymousAuthentication: nil,
                EventTTL: nil,
                Logging: nil,
                DefaultNotReadyTolerationSeconds: nil,
                DefaultUnreachableTolerationSeconds: nil,
                EncryptionConfig: nil,
            },
            KubeControllerManager: nil,
            KubeScheduler: nil,
            KubeProxy: nil,
            Kubelet: nil,
            Version: "1.28.6",
            VerticalPodAutoscaler: nil,
            EnableStaticTokenKubeconfig: false,
        },
        Networking: {
            Type: "calico",
            ProviderConfig: nil,
            Pods: "10.128.0.0/12",
            Nodes: "10.250.0.0/19",
            Services: "10.112.0.0/16",
            IPFamilies: nil,
        },
        Maintenance: {
            AutoUpdate: {
                KubernetesVersion: true,
                MachineImageVersion: true,
            },
            TimeWindow: nil,
            ConfineSpecUpdateRollout: nil,
        },
        Monitoring: nil,
        Provider: {
            Type: "azure",
            ControlPlaneConfig: {
                Raw: "{\"apiVersion\":\"azure.provider.extensions.gardener.cloud/v1alpha1\",\"kind\":\"ControlPlaneConfig\"}",
                Object: nil,
            },
            InfrastructureConfig: {
                Raw: "{\"apiVersion\":\"azure.provider.extensions.gardener.cloud/v1alpha1\",\"kind\":\"InfrastructureConfig\",\"networks\":{\"vnet\":{\"cidr\":\"10.250.0.0/19\"},\"zones\":[{\"cidr\":\"10.250.0.0/24\",\"name\":1,\"natGateway\":{\"enabled\":true,\"idleConnectionTimeoutMinutes\":4}},{\"cidr\":\"10.250.1.0/20\",\"name\":2,\"natGateway\":{\"enabled\":true,\"idleConnectionTimeoutMinutes\":4}}]},\"zoned\":true}",
                Object: nil,
            },
            Workers: [
                {
                    Annotations: nil,
                    CABundle: nil,
                    CRI: nil,
                    Kubernetes: nil,
                    Labels: nil,
                    Name: "cpu-worker-0",
                    Machine: {
                        Type: "Standard_D4_v3",
                        Image: nil,
                        Architecture: nil,
                    },
                    Maximum: 4,
                    Minimum: 2,
                    MaxSurge: {Type: 0, IntVal: 4, StrVal: ""},
        ...

Gomega truncated this representation as it exceeds 'format.MaxLength'.
Consider having the object provide a custom 'GomegaStringer' representation
or adjust the parameters in Gomega's 'format' package.

Learn more here: https://onsi.github.io/gomega/#adjusting-output

to equal
    <v1beta1.ShootSpec>: {
        Addons: nil,
        CloudProfileName: "az",
        DNS: nil,
        Extensions: [
            {
                Type: "shoot-dns-service",
                ProviderConfig: {
                    Raw: "{\"apiVersion\":\"service.dns.extensions.gardener.cloud/v1alpha1\",\"dnsProviderReplication\":{\"enabled\":true},\"kind\":\"DNSConfig\"}",
                    Object: nil,
                },
                Disabled: nil,
            },
            {
                Type: "shoot-cert-service",
                ProviderConfig: {
                    Raw: "{\"apiVersion\":\"service.cert.extensions.gardener.cloud/v1alpha1\",\"kind\":\"CertConfig\",\"shootIssuers\":{\"enabled\":true}}",
                    Object: nil,
                },
                Disabled: nil,
            },
            {
                Type: "shoot-networking-filter",
                ProviderConfig: nil,
                Disabled: true,
            },
        ],
        Hibernation: nil,
        Kubernetes: {
            ClusterAutoscaler: nil,
            KubeAPIServer: {
                KubernetesConfig: {FeatureGates: nil},
                AdmissionPlugins: nil,
                APIAudiences: nil,
                AuditConfig: {
                    AuditPolicy: {
                        ConfigMapRef: {Kind: "", Namespace: "", Name: "audit-policy", UID: "", APIVersion: "", ResourceVersion: "", FieldPath: ""},
                    },
                },
                OIDCConfig: nil,
                RuntimeConfig: nil,
                ServiceAccountConfig: nil,
                WatchCacheSizes: nil,
                Requests: nil,
                EnableAnonymousAuthentication: nil,
                EventTTL: nil,
                Logging: nil,
                DefaultNotReadyTolerationSeconds: nil,
                DefaultUnreachableTolerationSeconds: nil,
                EncryptionConfig: nil,
            },
            KubeControllerManager: nil,
            KubeScheduler: nil,
            KubeProxy: nil,
            Kubelet: nil,
            Version: "1.28.6",
            VerticalPodAutoscaler: nil,
            EnableStaticTokenKubeconfig: false,
        },
        Networking: {
            Type: "calico",
            ProviderConfig: nil,
            Pods: "10.128.0.0/12",
            Nodes: "10.250.0.0/19",
            Services: "10.112.0.0/16",
            IPFamilies: nil,
        },
        Maintenance: {
            AutoUpdate: {
                KubernetesVersion: true,
                MachineImageVersion: true,
            },
            TimeWindow: nil,
            ConfineSpecUpdateRollout: nil,
        },
        Monitoring: nil,
        Provider: {
            Type: "azure",
            ControlPlaneConfig: {
                Raw: "{\"apiVersion\":\"azure.provider.extensions.gardener.cloud/v1alpha1\",\"kind\":\"ControlPlaneConfig\"}",
                Object: nil,
            },
            InfrastructureConfig: {
                Raw: "{\"apiVersion\":\"azure.provider.extensions.gardener.cloud/v1alpha1\",\"kind\":\"InfrastructureConfig\",\"networks\":{\"vnet\":{\"cidr\":\"10.250.0.0/19\"},\"zones\":[{\"cidr\":\"10.250.0.0/24\",\"name\":1,\"natGateway\":{\"enabled\":true,\"idleConnectionTimeoutMinutes\":4}},{\"cidr\":\"10.250.1.0/24\",\"name\":2,\"natGateway\":{\"enabled\":true,\"idleConnectionTimeoutMinutes\":4}}]},\"zoned\":true}",
                Object: nil,
            },
            Workers: [
                {
                    Annotations: nil,
                    CABundle: nil,
                    CRI: nil,
                    Kubernetes: nil,
                    Labels: nil,
                    Name: "cpu-worker-0",
                    Machine: {
                        Type: "Standard_D4_v3",
                        Image: nil,
                        Architecture: nil,
                    },
                    Maximum: 4,
                    Minimum: 2,
                    MaxSurge: {Type: 0, IntVal: 4, StrVal: ""},
                  ...

Gomega truncated this representation as it exceeds 'format.MaxLength'.
Consider having the object provide a custom 'GomegaStringer' representation
or adjust the parameters in Gomega's 'format' package.

Learn more here: https://onsi.github.io/gomega/#adjusting-output
```

**Comment by akgalwas** [comment]
> I have changed three fields
> 
> * `spec.CloudProfileName`
> * `spec.shootIssuer.enabled`
> * `spec.provider.infrastructureConfig.networks.cidr`
>   and compared the files.
> 
> I expected the output to say that the 3 fields are different and show the difference. Instead, the tool said that just `spec` are different and printed the whole content. It was also too big so it was truncated. Is it intentional?
> 
> ```
> go run main.go files ~/Desktop/garden-kyma-dev-md-pvc-prov5.yaml ~/Desktop/prov-modified.yaml
> Comparing files: /Users/i316752/Desktop/garden-kyma-dev-md-pvc-prov5.yaml and /Users/i316752/Desktop/prov-modified.yaml
> Shoot files are NOT equal
> spec: Expected
>     <v1beta1.ShootSpec>: {
>         Addons: nil,
>         CloudProfileName: "az-modified",
>         DNS: nil,
>         Extensions: [
>             {
>                 Type: "shoot-dns-service",
>                 ProviderConfig: {
>                     Raw: "{\"apiVersion\":\"service.dns.extensions.gardener.cloud/v1alpha1\",\"dnsProviderReplication\":{\"enabled\":true},\"kind\":\"DNSConfig\"}",
>                     Object: nil,
>                 },
>                 Disabled: nil,
>             },
>             {
>                 Type: "shoot-cert-service",
>                 ProviderConfig: {
>                     Raw: "{\"apiVersion\":\"service.cert.extensions.gardener.cloud/v1alpha1\",\"kind\":\"CertConfig\",\"shootIssuers\":{\"enabled\":false}}",
>                     Object: nil,
>                 },
>                 Disabled: nil,
>             },
>             {
>                 Type: "shoot-networking-filter",
>                 ProviderConfig: nil,
>                 Disabled: true,
>             },
>         ],
>         Hibernation: nil,
>         Kubernetes: {
>             ClusterAutoscaler: nil,
>             KubeAPIServer: {
>                 KubernetesConfig: {FeatureGates: nil},
>                 AdmissionPlugins: nil,
>                 APIAudiences: nil,
>                 AuditConfig: {
>                     AuditPolicy: {
>                         ConfigMapRef: {Kind: "", Namespace: "", Name: "audit-policy", UID: "", APIVersion: "", ResourceVersion: "", FieldPath: ""},
>                     },
>                 },
>                 OIDCConfig: nil,
>                 RuntimeConfig: nil,
>                 ServiceAccountConfig: nil,
>                 WatchCacheSizes: nil,
>                 Requests: nil,
>                 EnableAnonymousAuthentication: nil,
>                 EventTTL: nil,
>                 Logging: nil,
>                 DefaultNotReadyTolerationSeconds: nil,
>                 DefaultUnreachableTolerationSeconds: nil,
>                 EncryptionConfig: nil,
>             },
>             KubeControllerManager: nil,
>             KubeScheduler: nil,
>             KubeProxy: nil,
>             Kubelet: nil,
>             Version: "1.28.6",
>             VerticalPodAutoscaler: nil,
>             EnableStaticTokenKubeconfig: false,
>         },
>         Networking: {
>             Type: "calico",
>             ProviderConfig: nil,
>             Pods: "10.128.0.0/12",
>             Nodes: "10.250.0.0/19",
>             Services: "10.112.0.0/16",
>             IPFamilies: nil,
>         },
>         Maintenance: {
>             AutoUpdate: {
>                 KubernetesVersion: true,
>                 MachineImageVersion: true,
>             },
>             TimeWindow: nil,
>             ConfineSpecUpdateRollout: nil,
>         },
>         Monitoring: nil,
>         Provider: {
>             Type: "azure",
>             ControlPlaneConfig: {
>                 Raw: "{\"apiVersion\":\"azure.provider.extensions.gardener.cloud/v1alpha1\",\"kind\":\"ControlPlaneConfig\"}",
>                 Object: nil,
>             },
>             InfrastructureConfig: {
>                 Raw: "{\"apiVersion\":\"azure.provider.extensions.gardener.cloud/v1alpha1\",\"kind\":\"InfrastructureConfig\",\"networks\":{\"vnet\":{\"cidr\":\"10.250.0.0/19\"},\"zones\":[{\"cidr\":\"10.250.0.0/24\",\"name\":1,\"natGateway\":{\"enabled\":true,\"idleConnectionTimeoutMinutes\":4}},{\"cidr\":\"10.250.1.0/20\",\"name\":2,\"natGateway\":{\"enabled\":true,\"idleConnectionTimeoutMinutes\":4}}]},\"zoned\":true}",
>                 Object: nil,
>             },
>             Workers: [
>                 {
>                     Annotations: nil,
>                     CABundle: nil,
>                     CRI: nil,
>                     Kubernetes: nil,
>                     Labels: nil,
>                     Name: "cpu-worker-0",
>                     Machine: {
>                         Type: "Standard_D4_v3",
>                         Image: nil,
>                         Architecture: nil,
>                     },
>                     Maximum: 4,
>                     Minimum: 2,
>                     MaxSurge: {Type: 0, IntVal: 4, StrVal: ""},
>         ...
> 
> Gomega truncated this representation as it exceeds 'format.MaxLength'.
> Consider having the object provide a custom 'GomegaStringer' representation
> or adjust the parameters in Gomega's 'format' package.
> 
> Learn more here: https://onsi.github.io/gomega/#adjusting-output
> 
> to equal
>     <v1beta1.ShootSpec>: {
>         Addons: nil,
>         CloudProfileName: "az",
>         DNS: nil,
>         Extensions: [
>             {
>                 Type: "shoot-dns-service",
>                 ProviderConfig: {
>                     Raw: "{\"apiVersion\":\"service.dns.extensions.gardener.cloud/v1alpha1\",\"dnsProviderReplication\":{\"enabled\":true},\"kind\":\"DNSConfig\"}",
>                     Object: nil,
>                 },
>                 Disabled: nil,
>             },
>             {
>                 Type: "shoot-cert-service",
>                 ProviderConfig: {
>                     Raw: "{\"apiVersion\":\"service.cert.extensions.gardener.cloud/v1alpha1\",\"kind\":\"CertConfig\",\"shootIssuers\":{\"enabled\":true}}",
>                     Object: nil,
>                 },
>                 Disabled: nil,
>             },
>             {
>                 Type: "shoot-networking-filter",
>                 ProviderConfig: nil,
>                 Disabled: true,
>             },
>         ],
>         Hibernation: nil,
>         Kubernetes: {
>             ClusterAutoscaler: nil,
>             KubeAPIServer: {
>                 KubernetesConfig: {FeatureGates: nil},
>                 AdmissionPlugins: nil,
>                 APIAudiences: nil,
>                 AuditConfig: {
>                     AuditPolicy: {
>                         ConfigMapRef: {Kind: "", Namespace: "", Name: "audit-policy", UID: "", APIVersion: "", ResourceVersion: "", FieldPath: ""},
>                     },
>                 },
>                 OIDCConfig: nil,
>                 RuntimeConfig: nil,
>                 ServiceAccountConfig: nil,
>                 WatchCacheSizes: nil,
>                 Requests: nil,
>                 EnableAnonymousAuthentication: nil,
>                 EventTTL: nil,
>                 Logging: nil,
>                 DefaultNotReadyTolerationSeconds: nil,
>                 DefaultUnreachableTolerationSeconds: nil,
>                 EncryptionConfig: nil,
>             },
>             KubeControllerManager: nil,
>             KubeScheduler: nil,
>             KubeProxy: nil,
>             Kubelet: nil,
>             Version: "1.28.6",
>             VerticalPodAutoscaler: nil,
>             EnableStaticTokenKubeconfig: false,
>         },
>         Networking: {
>             Type: "calico",
>             ProviderConfig: nil,
>             Pods: "10.128.0.0/12",
>             Nodes: "10.250.0.0/19",
>             Services: "10.112.0.0/16",
>             IPFamilies: nil,
>         },
>         Maintenance: {
>             AutoUpdate: {
>                 KubernetesVersion: true,
>                 MachineImageVersion: true,
>             },
>             TimeWindow: nil,
>             ConfineSpecUpdateRollout: nil,
>         },
>         Monitoring: nil,
>         Provider: {
>             Type: "azure",
>             ControlPlaneConfig: {
>                 Raw: "{\"apiVersion\":\"azure.provider.extensions.gardener.cloud/v1alpha1\",\"kind\":\"ControlPlaneConfig\"}",
>                 Object: nil,
>             },
>             InfrastructureConfig: {
>                 Raw: "{\"apiVersion\":\"azure.provider.extensions.gardener.cloud/v1alpha1\",\"kind\":\"InfrastructureConfig\",\"networks\":{\"vnet\":{\"cidr\":\"10.250.0.0/19\"},\"zones\":[{\"cidr\":\"10.250.0.0/24\",\"name\":1,\"natGateway\":{\"enabled\":true,\"idleConnectionTimeoutMinutes\":4}},{\"cidr\":\"10.250.1.0/24\",\"name\":2,\"natGateway\":{\"enabled\":true,\"idleConnectionTimeoutMinutes\":4}}]},\"zoned\":true}",
>                 Object: nil,
>             },
>             Workers: [
>                 {
>                     Annotations: nil,
>                     CABundle: nil,
>                     CRI: nil,
>                     Kubernetes: nil,
>                     Labels: nil,
>                     Name: "cpu-worker-0",
>                     Machine: {
>                         Type: "Standard_D4_v3",
>                         Image: nil,
>                         Architecture: nil,
>                     },
>                     Maximum: 4,
>                     Minimum: 2,
>                     MaxSurge: {Type: 0, IntVal: 4, StrVal: ""},
>                   ...
> 
> Gomega truncated this representation as it exceeds 'format.MaxLength'.
> Consider having the object provide a custom 'GomegaStringer' representation
> or adjust the parameters in Gomega's 'format' package.
> 
> Learn more here: https://onsi.github.io/gomega/#adjusting-output
> ```

Yes it is. The current version uses matcher to compare `spec` object. In the future we may want to have it more granular (define separate matches for all objects that are under spec). I didn't changed the way how matchers work as the goal of this PR is to create a base for next steps.

**Comment by akgalwas** [comment]
> 1. @akgalwas have you considered placing the tool in the `hack` repository instead of `tools`?
> 2. Please create a README.md that will explain briefly the purpose of the tool and show an example usage.

1. Moved the code to `hack` folder
2. Created minimal README.md file. It will be enhanced later once we add more functionalities.

**Review by Disper — APPROVED** [review]

File: hack/shoot-comparator/README.md [hunk 1]
```diff
@@ -0,0 +1,20 @@
+# Shoot comparator
+
+## Overview
+
+This tool compares shoot files created by KIM, and Provisioner. It was created to ensure the KIM generates exactly the same shoots as Provisioner does.
```
Comment (grego952) [inline hack/shoot-comparator/README.md:5]:
```suggestion
This tool compares shoot files created by KIM and Provisioner. It was created to ensure that KIM generates exactly the same shoots as Provisioner does.
```

File: hack/shoot-comparator/README.md [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+- https://github.com/kyma-project/infrastructure-manager/issues/250
+
+## Build
+```
+CGO_ENABLED=0 GOOS=darwin GOARCH=arm64 go build -o ./bin/comparator ./cmd
+```
+
+## Run
+
+In order to compare files execute the following command:
```
Comment (grego952) [inline hack/shoot-comparator/README.md:17]:
```suggestion
To compare files, execute the following command:
```

**Review by Disper — APPROVED** [review]

**Review by grego952 — APPROVED** [review]

**Review by Disper — APPROVED** [review]

**Pull request merged**